File: first-round/494.目标和.js
/*
 * @lc app=leetcode.cn id=494 lang=javascript
 *
 * [494] 目标和
 *
 * https://leetcode-cn.com/problems/target-sum/description/
 *
 * algorithms
 * Medium (45.73%)
 * Likes:    661
 * Dislikes: 0
 * Total Accepted:    83.1K
 * Total Submissions: 181.8K
 * Testcase Example:  '[1,1,1,1,1]\n3'
 *
 * 给定一个非负整数数组，a1, a2, ..., an, 和一个目标数，S。现在你有两个符号 + 和 -。对于数组中的任意一个整数，你都可以从 + 或
 * -中选择一个符号添加在前面。
 * 
 * 返回可以使最终数组和为目标数 S 的所有添加符号的方法数。
 * 
 * 
 * 
 * 示例：
 * 
 * 输入：nums: [1, 1, 1, 1, 1], S: 3
 * 输出：5
 * 解释：
 * 
 * -1+1+1+1+1 = 3
 * +1-1+1+1+1 = 3
 * +1+1-1+1+1 = 3
 * +1+1+1-1+1 = 3
 * +1+1+1+1-1 = 3
 * 
 * 一共有5种方法让最终目标和为3。
 * 
 * 
 * 
 * 
 * 提示：
 * 
 * 
 * 数组非空，且长度不会超过 20 。
 * 初始的数组的和不会超过 1000 。
 * 保证返回的最终结果能被 32 位整数存下。
 * 
 * 
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var findTargetSumWays = function(nums, target) {
  let prev = []

  for (let i = 0; i < nums.length; i++) {
    const curr = []

    for (let j = -1000; j <= 1000; j++) {
      if (i === 0) {
        if (j === 0) curr[j] = nums[i] === j ? 2 : 0
        else curr[j] = nums[i] === j || nums[i] === -1 * j ? 1 : 0
      } else {
        curr[j] = (prev[j - nums[i]] || 0) + (prev[j + nums[i]] || 0)
      }
    }

    prev = curr
  }

  return prev[target]
};
// @lc code=end

File: third-round/494.目标和.py
#
# @lc app=leetcode.cn id=494 lang=python3
#
# [494] 目标和
#
# https://leetcode-cn.com/problems/target-sum/description/
#
# algorithms
# Medium (49.09%)
# Likes:    1125
# Dislikes: 0
# Total Accepted:    199.1K
# Total Submissions: 405.7K
# Testcase Example:  '[1,1,1,1,1]\n3'
#
# 给你一个整数数组 nums 和一个整数 target 。
# 
# 向数组中的每个整数前添加 '+' 或 '-' ，然后串联起所有整数，可以构造一个 表达式 ：
# 
# 
# 例如，nums = [2, 1] ，可以在 2 之前添加 '+' ，在 1 之前添加 '-' ，然后串联起来得到表达式 "+2-1" 。
# 
# 
# 返回可以通过上述方法构造的、运算结果等于 target 的不同 表达式 的数目。
# 
# 
# 
# 示例 1：
# 
# 
# 输入：nums = [1,1,1,1,1], target = 3
# 输出：5
# 解释：一共有 5 种方法让最终目标和为 3 。
# -1 + 1 + 1 + 1 + 1 = 3
# +1 - 1 + 1 + 1 + 1 = 3
# +1 + 1 - 1 + 1 + 1 = 3
# +1 + 1 + 1 - 1 + 1 = 3
# +1 + 1 + 1 + 1 - 1 = 3
# 
# 
# 示例 2：
# 
# 
# 输入：nums = [1], target = 1
# 输出：1
# 
# 
# 
# 
# 提示：
# 
# 
# 1 
# 0 
# 0 
# -1000 
# 
# 
#

# @lc code=start
from collections import defaultdict


class Solution:
    def findTargetSumWays(self, nums: list[int], target: int) -> int:
        sum_n = sum(nums)

        if not -sum_n <= target <= sum_n:
            return 0

        dp = [defaultdict(int) for _ in range(len(nums) + 1)]
        dp[0][0] = 1

        for i in range(1, len(nums) + 1):
            n = nums[i - 1]

            for j in range(-sum_n, sum_n + 1):
                dp[i][j] = dp[i - 1][j - n] + dp[i - 1][j + n]

        return dp[-1][target]
# @lc code=end

s = Solution()
print(s.findTargetSumWays([1,1,1,1,1], target = 3))
print(s.findTargetSumWays([1,1,0,1,1], target = 4))
print(s.findTargetSumWays([1], target = 1))
  